packages/transformers/js/core/src/collect.rs

```diff
@@ -0,0 +1,1033 @@
+            }
+          }
+        }
+      }
+      _ => {
+        // Non-static.
+        self.non_static_requires.insert(src.clone());
+        let span = match node {
+          Pat::Ident(id) => id.id.span,
+          Pat::Array(arr) => arr.span,
+          Pat::Object(obj) => obj.span,
+          Pat::Rest(rest) => rest.span,
+          Pat::Assign(assign) => assign.span,
+          Pat::Invalid(i) => i.span,
+          Pat::Expr(_) => DUMMY_SP,
+        };
+        self.add_bailout(span, BailoutReason::NonStaticDestructuring);
+      }
+    }
+  }
+
+  pub fn get_non_const_binding_idents(&self, node: &Pat, idents: &mut Vec<Ident>) {
+    match node {
+      Pat::Ident(ident) => {
+        if self.non_const_bindings.contains_key(&id!(ident.id)) {
+          idents.push(ident.id.clone());
+        }
+      }
+      Pat::Object(object) => {
+        for prop in &object.props {
+          match prop {
+            ObjectPatProp::KeyValue(kv) => {
+              self.get_non_const_binding_idents(&*kv.value, idents);
+            }
+            ObjectPatProp::Assign(assign) => {
+              if self.non_const_bindings.contains_key(&id!(assign.key)) {
+                idents.push(assign.key.clone());
+              }
+            }
+            ObjectPatProp::Rest(rest) => {
+              self.get_non_const_binding_idents(&*rest.arg, idents);
+            }
+          }
+        }
+      }
+      Pat::Array(array) => {
+        for el in array.elems.iter().flatten() {
+          self.get_non_const_binding_idents(el, idents);
+        }
+      }
+      _ => {}
+    }
+  }
+
+  fn add_bailout(&mut self, span: Span, reason: BailoutReason) {
+    if let Some(bailouts) = &mut self.bailouts {
+      bailouts.push(Bailout {
+        loc: SourceLocation::from(&self.source_map, span),
+        reason,
+      })
+    }
+  }
+}
+
+fn has_binding_identifier(node: &Pat, sym: &JsWord, decls: &HashSet<IdentId>) -> bool {
+  match node {
+    Pat::Ident(ident) => {
+      if ident.id.sym == *sym && !decls.contains(&id!(ident.id)) {
+        return true;
+      }
+    }
+    Pat::Object(object) => {
+      for prop in &object.props {
+        match prop {
+          ObjectPatProp::KeyValue(kv) => {
+            if has_binding_identifier(&*kv.value, sym, decls) {
+              return true;
+            }
+          }
+          ObjectPatProp::Assign(assign) => {
+            if assign.key.sym == *sym && !decls.contains(&id!(assign.key)) {
+              return true;
+            }
+          }
+          ObjectPatProp::Rest(rest) => {
+            if has_binding_identifier(&*rest.arg, sym, decls) {
+              return true;
+            }
+          }
+        }
+      }
+    }
+    Pat::Array(array) => {
+      for el in array.elems.iter().flatten() {
+        if has_binding_identifier(el, sym, decls) {
+          return true;
+        }
+      }
+    }
+    _ => {}
+  }
+
+  false
+}
```

Comment:
Kinda hard to see what changed since this moved... were there changes or is it pretty much the same?

Reply:
Good point, I've moved this back into hoist.rs so that you can see what changed.
There were a few changes, mainly because previously reexports weren't collected and for exports the mapping `local variable -> exported name` provides enough information for the hoisting, but not for general purpose symbols information. (Now there is a `exported name -> local variable` mapping and then also a `exported name -> last local variable` that behaves as before.)